File: Day7/solution.py
f = open('input.txt', 'r')
ls = [line.replace('bags', 'bag').split('bag')
      for line in f.read().splitlines()]
ls = [list(filter(lambda s: len(s) != 1, line)) for line in ls]
ls = [list(map(lambda s: s.replace(' contain', '').strip(',. '), line))
      for line in ls]




c_in = {line[0]: {bag[2:] for bag in line[1:]} for line in ls}
c_by ={line[0]: {key for key, value in c_in.items() if line[0] in value} for line in ls}
print(c_by)


contains_gold = set()

def check(colour):
    for cl in c_by[colour]:
        contains_gold.add(cl)
        check(cl)


check('shiny gold')
print(len(contains_gold))

File: Day7/solution.hs
import Data.List (isInfixOf)
import Data.Map (Map)
import Data.Graph (Graph, Vertex, graphFromEdges, graphFromEdges', transposeG, dfs)
import Data.List.Split
import Text.Read
import Data.Maybe (isNothing)
import GHC.Unicode (isDigit, isSpace)

main = do
    input <- readFile "test.txt"
    let sLines = map (removeWhiteSpace . splitOn "  " . removeNumbers . removeWordContaining "bag" . removeWordContaining "contain") $ splitOn "\n" input
    let (g, v, k) = graphFromEdges ([(head line, head line, tail line) | line <- sLines])
    let gT = transposeG g
    print $ v 3
    print $ length $ dfs gT [7] 

removeWordContaining:: String -> String -> String  
removeWordContaining target s = unwords $ map (\n -> if target `isInfixOf` n then "" else n ) $ words s

removeNumbers:: String -> String 
removeNumbers = filter (not. isDigit) 

pairList:: [String] -> [String]
pairList [] = []
pairList (x:y:xs) = (x++y): pairList xs

removeWhiteSpace:: [String] -> [String] 
removeWhiteSpace  = map $ filter $ not . isSpace  